### PR TITLE
Fixes #18781: Replace stats API by a prometheus endpoint

### PR DESCRIPTION
**Pull request by amousset**
https://issues.rudder.io/issues/18781

### CODE REVIEW

**Comment by amousset** [comment]
Still missing some docs for the metrics API

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

File: relay/sources/relayd/tests/processing_reporting.rs [hunk 1]
```diff
@@ -21,6 +21,37 @@ pub fn db_connection() -> PgConnection {
     PgConnection::establish("postgres://rudderreports:PASSWORD@127.0.0.1/rudder").unwrap()
 }
 
+/// List of key values to check in the metrics text
+pub fn check_prometheus(metrics: &str, mut expected: HashMap<&str, &str>) -> bool {
+    let mut is_ok = true;
+    println!("{}", metrics);
```
Comment (gpoblon) [inline relay/sources/relayd/tests/processing_reporting.rs:27]: not so sure, is this a debug print ?

Comment (amousset) [reply relay/sources/relayd/tests/processing_reporting.rs:27]: indeed, should not be necessary anymore

File: relay/sources/relayd/src/api/metrics.rs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        if let Err(e) = encoder.encode(&REGISTRY.gather(), &mut buffer) {
+            return Err(reject::custom(RudderReject::new(e)));
+        };
+        let mut res = match String::from_utf8(buffer.clone()) {
+            Ok(v) => v,
+            Err(e) => return Err(reject::custom(RudderReject::new(e))),
+        };
+        buffer.clear();
+
+        let mut buffer = Vec::new();
```
Comment (gpoblon) [inline relay/sources/relayd/src/api/metrics.rs:33]: since it has just been cleared I guess you dont need to declare buffer again

Comment (gpoblon) [reply relay/sources/relayd/src/api/metrics.rs:33]: You could also define a closure (called twice) to do the encoding

**Review by gpoblon — APPROVED** [review]

File: relay/sources/relayd/src/api/metrics.rs (hunk 2 above)
Comment (amousset) [reply relay/sources/relayd/src/api/metrics.rs:33]: I have to admit I just copy-pasted the lib example here, but it can be factored easily

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

File: relay/sources/relayd/src/api/metrics.rs (hunk 2 above)
Comment (amousset) [reply relay/sources/relayd/src/api/metrics.rs:23]: A closure is indeed way better!

```rust
    pub async fn metrics() -> Result<impl Reply, Rejection> {
        use prometheus::Encoder;
        let encoder = prometheus::TextEncoder::new();
        let mut buffer = Vec::new();
        let mut encode = |metrics: &[MetricFamily]| match encoder.encode(&metrics, &mut buffer) {
            Ok(_) => Ok(()),
            Err(e) => Err(reject::custom(RudderReject::new(e))),
        };

        encode(&REGISTRY.gather())?;
        encode(&prometheus::gather())?;
        Ok(buffer)
    }
```

**Comment by amousset** [comment]
OK, squash merging this PR

**Pull request merged**